### ISSUE TITLE
Is WPF going to die? Why TextBox still has not placeholder/watermark?

### ISSUE THREAD

**Issue by GF-Huang** [dotnet/wpf#4345]
Such a basic feature, after all these years, has still not been added to WPF.

**Comment by lindexi** [comment]
@GF-Huang Do not do that. This repository is a WPF framework, which is a low-level framework. This repo should only provide a basic framework, not too much business logic and business controls. You can use other professional WPF control libraries to get more customized controls, such as your watermark.

See

- [HandyOrg/HandyControl: Contains some simple and commonly used WPF controls](https://github.com/HandyOrg/HandyControl )
- [MahApps/MahApps.Metro: A toolkit for creating Metro / Modern UI styled WPF apps.](https://github.com/MahApps/MahApps.Metro )
- [Panuon/PanuonUI.Silver: Panuon.UI optimized version. A beautiful wpf ui library using templates & attached properties.](https://github.com/Panuon/PanuonUI.Silver )
- [spiegelp/MaterialDesignExtensions: Material Design Extensions is based on Material Design in XAML Toolkit to provide additional controls and features for WPF apps](https://github.com/spiegelp/MaterialDesignExtensions )
- [LeoYang-Chuese/wif: WPF Infrastructure Framework（WIF）is an infrastructure framework that helps you develop WPF applications quickly.](https://github.com/LeoYang-Chuese/wif )
- [lvaleriu/Virtualization: Data virtualization in WPF, UI virtualization - Credits go to Bea Stollnitz, Paul McClean and Vincent Van Den Berghe for their implementations](https://github.com/lvaleriu/Virtualization )
- [jiaze.li/RadarControl WPF](https://gitee.com/jiaze.li/RadarControl )
- [a44281071/SunTaiLibrary](https://github.com/a44281071/SunTaiLibrary )
- [quicoli/WPF-AutoComplete-TextBox: An autocomplete TextBox for WPF](https://github.com/quicoli/WPF-AutoComplete-TextBox )
- [ratishphilip/wpfspark: A rich UserControl library to enhance the look and feel of WPF applications.](https://github.com/ratishphilip/wpfspark )
- [rulyotano/wpf-drag-animated-panel: Panel for WPF that re-order elements when dragging items.](https://github.com/rulyotano/wpf-drag-animated-panel )
- [firstfloorsoftware/mui: Modern UI for WPF](https://github.com/firstfloorsoftware/mui )
- [Dirkster99/SuggestBoxLib: A WPF Dark/Light AutoComplete TextBox that can easily handle 20.000+ entries.](https://github.com/Dirkster99/SuggestBoxLib )
- [microsoft/InteractiveDataDisplay.WPF: Interactive Data Display for WPF is a set of controls for adding interactive visualization of dynamic data to your application. It allows to create line graphs, bubble charts, heat maps and other complex 2D plots which are very common in scientific software. Interactive Data Display for WPF integrates well with Bing Maps control to show data on a geographic map in latitude/longitude coordinates. The controls can also be operated programmatically.](https://github.com/microsoft/InteractiveDataDisplay.WPF )
- [sketchpunk/openkeyboard: Virtual keyboard with dynamic user defined layouts](https://github.com/sketchpunk/openkeyboard )

**Comment by GF-Huang** [comment]
Oh, it turns out that WPF is just an engine, not a car.

**Comment by Symbai** [comment]
> Do not do that. This repository is a WPF framework, which is a low-level framework. This repo should only provide a basic framework, not too much business logic and business controls. You can use other professional WPF control libraries to get more customized controls, such as your watermark.

First this is a dupe of #1127 and second your statement is not a good argument, as Winforms "recently" (after being open sourced) added this and winforms is older than WPF and Winforms team didn't told everyone to use professional Winforms control libraries. That means its a basic feature and should get supported natively. Again, see #1127

**Comment by wstaelens** [comment]
> > Do not do that. This repository is a WPF framework, which is a low-level framework. This repo should only provide a basic framework, not too much business logic and business controls. You can use other professional WPF control libraries to get more customized controls, such as your watermark.
> 
> First this is a dupe of #1127 and second your statement is not a good argument, as Winforms "recently" (after being open sourced) added this and winforms is older than WPF and Winforms team didn't told everyone to use professional Winforms control libraries. That means its a basic feature and should get supported natively. Again, see #1127

I agree Symbai. After all these years we were being pushed away from winforms. Now we are at WPF and years later the team still doesn't do a thing.

We only see them trying to move us to something else, which is even more broken and alpha/beta/whatever status: MAUI.
(see https://github.com/microsoft/microsoft-ui-xaml/issues/4650 )

**Comment by GF-Huang** [comment]
They just don't want to admit this fact and find a decent excuse to fool you.

**Comment by mjeson** [comment]
@GF-Huang  I found this textbox with bindable placeholder. May fit your need.

https://github.com/GuOrg/Gu.Wpf.Adorners
